File: android/beaconsdk/src/main/java/com/xenber/beaconsdk/core/BeaconLifecycle.kt
/** Check App Lifecycle For Beacon Scanning (FG/BG) **/

package com.xenber.beaconsdk.core

import android.app.Application
import android.content.Context
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import com.xenber.beaconsdk.util.Logger

class BeaconLifecycle (
  context: Context
) {

  private var isForeground: Boolean = true
  private var listener: LifecycleListener? = null
  
  init {
    android.os.Handler(android.os.Looper.getMainLooper()).post {
      ProcessLifecycleOwner
        .get()
        .lifecycle
        .addObserver(object: LifecycleEventObserver {

          override fun onStateChanged(
            source: LifecycleOwner,
            event: Lifecycle.Event
          ) {
            when (event) {
              Lifecycle.Event.ON_START -> {
                isForeground = true
                Logger.i("App moved to FOREGROUND")
                listener?.onForeground()
              }
              Lifecycle.Event.ON_STOP -> {
                isForeground = false
                Logger.i("App moved to BACKGROUND")
                listener?.onBackground()
              }
              else -> Unit
            }
          }
        })
      
        Logger.i("BeaconLifecycle observer registered")
      }
    }

    fun isForeground(): Boolean = isForeground

    fun setListener(
      listener : LifecycleListener
    ){
      this.listener = listener
    }

    interface LifecycleListener  {
      fun onForeground()
      fun onBackground()
    }
}

File: ios/BeaconSdk/Core/BeaconLifecycle.swift
import UIKit
import Foundation

protocol LifecycleListener: AnyObject {
  func onForeground()
  func onBackground()
}

class BeaconLifecycle {

  private var isForegroundState: Bool = true
  private weak var listener: LifecycleListener?

  init() {
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(didEnterBackground),
      name: UIApplication.didEnterBackgroundNotification,
      object: nil
    )
    
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(didBecomeActive),
      name: UIApplication.didBecomeActiveNotification,
      object: nil
    )
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  var isForeground: Bool {
    return isForegroundState
  }

  func setListener(_ listener: LifecycleListener){
    self.listener = listener
  }

  @objc private func didBecomeActive() {
    isForegroundState = true
    Logger.i("App moved to FOREGROUND")
    listener?.onForeground()
  }
    
  @objc private func didEnterBackground() {
    isForegroundState = false
    Logger.i("App moved to BACKGROUND")
    listener?.onBackground()
  }
}